src/network/network.cpp

```diff
@@ -7,201 +7,219 @@
+  host_->registerCapability(taraxa_capability_);
 }
 
-void Network::stop() {
+Network::~Network() {
   if (bool b = false; !stopped_.compare_exchange_strong(b, !b)) {
     return;
   }
-
+  tp_.stop();
   host_->stop();
   if (!network_file_.empty()) {
-    saveNetwork(network_file_);
+    if (auto netData = host_->saveNetwork(); !netData.empty()) {
+      writeFile(network_file_.string(), &netData);
+      LOG(log_dg_) << "Network saved to: " << network_file_;
+    }
   }
-  taraxa_capability_->onStopping();
 }
 
-void Network::sendTest(NodeID const &id) {
-  taraxa_capability_->sendTestMessage(id, 1);
-  LOG(log_dg_) << "Sent test";
+void Network::start() {
+  if (bool b = true; !stopped_.compare_exchange_strong(b, !b)) {
+    return;
+  }
+  assert(!host_->isStarted());
+  host_->start();
+  if (!boot_nodes_.empty()) {
+    for (auto const &[k, v] : boot_nodes_) {
+      host_->addNode(k, v);
+    }
+    // Every 30 seconds check if connected to another node and refresh boot nodes
+    util::post_periodic(tp_.unsafe_get_io_context(), {30000}, [this] {
+      // If node count drops to zero add boot nodes again and retry
+      if (getNodeCount() == 0) {
+        for (auto const &[k, v] : boot_nodes_) {
+          host_->addNode(k, v);
+        }
+      }
+      if (host_->peerCount() == 0) {
+        for (auto const &[k, _] : boot_nodes_) {
+          host_->invalidateNode(k);
+        }
+      }
+    });
+  }
+  tp_.start();
+  LOG(log_nf_) << "Started Network address: " << conf_.network_address << ":" << conf_.network_tcp_port << std::endl;
+  LOG(log_nf_) << "Started Node id: " << host_->id();
 }
 
-void Network::sendBlock(NodeID const &id, DagBlock const &blk) {
-  taraxa_capability_->sendBlock(id, blk);
-  LOG(log_dg_) << "Sent Block:" << blk.getHash().toString();
+bool Network::isStarted() { return !stopped_; }
+
+std::list<NodeEntry> Network::getAllNodes() const { return host_->getNodes(); }
+
+unsigned Network::getPeerCount() { return taraxa_capability_->getPeersCount(); }
+
+unsigned Network::getNodeCount() { return host_->getNodeCount(); }
+
+Json::Value Network::getStatus() { return taraxa_capability_->getStatus(); }
+
+std::vector<NodeID> Network::getAllPeers() const { return taraxa_capability_->getAllPeers(); }
+
+void Network::onNewBlockVerified(shared_ptr<DagBlock> const &blk) {
+  tp_.post([=] {
+    taraxa_capability_->onNewBlockVerified(*blk);
+    LOG(log_dg_) << "On new block verified:" << blk->getHash().toString();
+  });
 }
 
-void Network::sendTransactions(NodeID const &id, std::vector<taraxa::bytes> const &transactions) {
-  taraxa_capability_->sendTransactions(id, transactions);
-  LOG(log_dg_) << "Sent transactions:" << transactions.size();
+void Network::onNewTransactions(std::vector<taraxa::bytes> transactions) {
+  tp_.post([=, transactions = std::move(transactions)] {
+    taraxa_capability_->onNewTransactions(transactions, true);
```

Comment:
all  uitl::post_periodic and ba::post calls inside tarxa capability might be replaced by   tp_.pos and  tp_.pos_periodic